### PR TITLE
 Better detect BT devices and LTE devices

### PR DESCRIPTION
**Pull request by posterzh**
**Why**
Detect the BLE and LET devices in better way.

**How**
Get the BLE devices and LTE devices programatically using dbus.

**References**
https://github.com/NebraLtd/hm-diag/issues/96 
https://github.com/NebraLtd/hm-diag/issues/88

### CODE REVIEW

File: Dockerfile
Comment (shawaj) [inline Dockerfile:32]: IIRC @marvinmarnold and @vpetersson mentioned previously that there was little point in using venv in containerised setup.

Think was in relation to hm-config but guess it applies here too

File: Dockerfile
Comment (shawaj) [inline Dockerfile:47]: Do we need bluez and network manager in the container if we are communicating with them over dbus?

File: Dockerfile
Comment (shawaj) [inline Dockerfile:23]: IIRC @vpetersson thought it best to unpin packages previously

File: Dockerfile
Comment (marvinmarnold) [reply Dockerfile:32]: I still have a soft spot for venv since it makes python portable. But yea, we did agree to stop using it.

@posterzh hm-pktfwd is an example of the new pattern: https://github.com/NebraLtd/hm-pktfwd/blob/e067bef1be17f9d3c1036cbe1100b317d5cc5f57/Dockerfile#L65

File: Dockerfile
Comment (posterzh) [reply Dockerfile:21]: Make sense

File: Dockerfile
Comment (posterzh) [reply Dockerfile:46]: I will check.

**Comment by vpetersson** [comment]
Nice improvement, @posterzh!

File: Dockerfile
Comment (posterzh) [reply Dockerfile]: Nice pattern.

File: Dockerfile
Comment (posterzh) [reply Dockerfile]: @shawaj Without the dependencies of `bluez`, `network-manager`, `modem-manager`, detecting BT, WIFI, LTE devices works well with only `dbus` dependency. I removed unnecessary dependencies in the Dockerfile.

**Comment by posterzh** [comment]
I am going to migrate dbus python library from `dbus-python` to `pydbus`. 
- `dbus-python` is deprecated(https://github.com/posborne/dbus-python).
image
- To install `dbus-python`. we need to install `gcc` and `python-dev` package. multistage docker build is used for optimization for the size of container.
- `pydbus`(https://github.com/LEW21/pydbus) are installed well without above dependency in `raspberry-pi-debian-python` system.

File: Dockerfile
Comment (shawaj) [reply Dockerfile]: @marvinmarnold I am not fussed either way! Whichever you think is best TBH

**Comment by posterzh** [comment]
**Detecting BT and LTE devices using dbus:**

image

**Diagnostic Web page:**

image

**Diagnostic container from the multistage docker build:**

image

**Comment by shawaj** [comment]
Looks good to me, but probably worth @vpetersson or @marvinmarnold or someone else from @NebraLtd/developers taking a deeper look.

**Comment by vpetersson** [comment]
Good job @posterzh. This is a big improvement. I wonder however if we should move the detection logic to hm-pyhelper` instead as we should probably re-use this code in hm-diag too. Wdyt @marvinmarnold?

**Comment by brianwaller063** [comment]
> Good job @posterzh. This is a big improvement. I wonder however if we should move the detection logic to hm-pyhelper` instead as we should probably re-use this code in hm-diag too. Wdyt @marvinmarnold?

FYI, in the `hm-diag`, the only feature used is to check the presence of BT/LTE devices. Furthermore, I added the utility methods to get the details of all BT, WiFi, LTE(`get_bt_devices()`, `get_wifi_devices()`, `get_lte_devices()`). These details of devices are used in checking the presence for validating their state. Meanwhile, almost the same code can be used for managing and connecting WiFi devices.

**Comment by posterzh** [comment]
> Good job @posterzh. This is a big improvement. I wonder however if we should move the detection logic to hm-pyhelper` instead as we should probably re-use this code in hm-diag too. Wdyt @marvinmarnold?

FYI, in the `hm-diag`, the only feature used is to check the presence of BT/LTE devices. Furthermore, I added the utility methods to get the details of all BT, WiFi, LTE(`get_bt_devices()`, `get_wifi_devices()`, `get_lte_devices()`). These details of devices are used in checking the presence for validating their state. Meanwhile, almost the same code can be used for managing and connecting WiFi devices.

File: hw_diag/utilities/hardware.py
Comment (marvinmarnold) [inline hw_diag/utilities/hardware.py:8]: Can you use get_logger from hm_pyhelper?

File: hw_diag/utilities/hardware.py
Comment (marvinmarnold) [inline hw_diag/utilities/hardware.py:265]: This should be log instead of print.

**Review by marvinmarnold — CHANGES_REQUESTED** [review]
This work looks really good. The code is very readable and clear.

This logic also needs to be added to https://github.com/NebraLtd/hm-diag/blob/00c716dd4980e35ac7436d9f59aebb17bec07535/hw_diag/diagnostics/bt_lte_diagnostic.py Until we have completed https://github.com/NebraLtd/Hotspot-Production-Tool/issues/28 we will continue to have parallel logic for /json and /initFile.txt.

Lastly, my typical comment. Could you please add some basic unit tests?

File: hw_diag/utilities/hardware.py
Comment (posterzh) [reply hw_diag/utilities/hardware.py:8]: Ok, make sense.

**Review by vpetersson — CHANGES_REQUESTED** [review]

File: hw_diag/utilities/hardware.py
Comment (posterzh) [reply hw_diag/utilities/hardware.py:265]: Ok

File: hw_diag/utilities/hardware.py
Comment (posterzh) [reply hw_diag/utilities/hardware.py:101]: I agree. I will keep in mind it.

File: Dockerfile
Comment (posterzh) [reply Dockerfile:48]: Yes, the size of the docker image is same after removing the clean up.

**Comment by vpetersson** [comment]
@posterzh Please fix the linting errors.

File: Dockerfile
Comment (vpetersson) [inline Dockerfile:18]: Nitpicking here, but I'd probably run this on a separate `RUN` line just to help with debugging if for whatever reason it would fail building.

**Review by vpetersson — APPROVED** [review]
Approved by me, but I would ideally like to see some unit tests here. Don't merge this before @marvinmarnold approves it.

File: Dockerfile
Comment (posterzh) [reply Dockerfile:18]: It would help debugging.

Comment (posterzh) [reply Dockerfile:18]: > Approved by me, but I would ideally like to see some unit tests here. Don't merge this before @marvinmarnold approves it.

Yeah, I am writing the unit test and a parallel logic in `DiagnosticsReport`. These pushes are just for `In-progress`.

Comment (vpetersson) [reply Dockerfile]: I see. Please update the PR to 'Draft' mode if that is the case.

**Comment by shawaj** [comment]
> Good job @posterzh. This is a big improvement. I wonder however if we should move the detection logic to hm-pyhelper` instead as we should probably re-use this code in hm-diag too. Wdyt @marvinmarnold?

@vpetersson do you think we would need this in hm-config? We seem to be using some custom nmcli in there so not sure it'll be compatible https://github.com/NebraLtd/hm-config/blob/3b2e0fe138ab04533ffa44afa3d0a60d7972430e/gatewayconfig/processors/wifi_processor.py#L20

**Comment by posterzh** [comment]
> This work looks really good. The code is very readable and clear.
> 
> This logic also needs to be added to https://github.com/NebraLtd/hm-diag/blob/00c716dd4980e35ac7436d9f59aebb17bec07535/hw_diag/diagnostics/bt_lte_diagnostic.py Until we have completed [NebraLtd/Hotspot-Production-Tool#28](https://github.com/NebraLtd/Hotspot-Production-Tool/issues/28) we will continue to have parallel logic for /json and /initFile.txt.
> 
> Lastly, my typical comment. Could you please add some basic unit tests?

@marvinmarnold 
Unit tests are added.
What about using a separate `BtDiagnostics` and `LteDiagnostics` rather than `BtLteDiagnostics`?
https://github.com/NebraLtd/hm-diag/blob/c7253186e983d09e336f58fcefe5b69168e12c7d/hw_diag/diagnostics/bt_lte_diagnostic.py#L60

**Review by marvinmarnold — APPROVED** [review]
> What about using a separate BtDiagnostics and LteDiagnostics rather than BtLteDiagnostics?

Good idea but lets not let that hold things up. Merge this and create a follow up ticket  to do in a future sprint.

**Comment by posterzh** [comment]
@marvinmarnold I created a follow up ticket as above.

**Pull request merged**
